README.md

```diff
@@ -1 +1,279 @@
+And second, partitioning the work by time means the key space is replicated
+over each node making the tables inefficient and larger than they need to be.
+
+A solution to these problems is, firstly, to fork a new worker only after the
+aggregation table gets too big (thus providing scale-out memory along with
+scale-out throughput) and, secondly, to implement a shuffle to
+route the table rows to the worker that owns the row.  Ownership of rows
+is handled like spark and map-reduce using a hash partition.  The
+number of dynamic nodes is limited either because the cluster
+manager (e.g., kubernetes) indicates no more resources are available
+or a configured scale-out limit is hit.  Either way, when there are no
+more available resources, workers then spill to disk.
+
+In this approach, we begin with N=M workers for throughput parallelism then
+dynamically add workers (increasing N) to provide additional memory scale-out
+and throughput parallelism.  The traversal range is divided among the M workers
+in some way (i.e., evenly divided, partitioned with a block/M stride, etc)
+
+When a worker hits a memory limit because its aggregation table gets to big,
+it splits its job, say in 2 (though this fanout can be configured), or it resorts
+to spilling if no resource is available for the split.  When it splits,
+half of the key space and half of the input partition goes to the new child.
+
+Once split, the worker initiate a _continuous shuffle_, where each worker pushes
+rows that it does not own to the owner of those rows.  At any point in time,
+every worker knows the other N-1 workers in the job, how the keys are partitioned
+across the workers, and its individual position in the set of workers.  Thus,
+each worker knows where to send rows that it does not own.
+
+To optimize the continuous shuffle, each worker keeps N-1 _peer tables_ to batch
+the streaming updates.  As records arrive in this table, they can be aggregated
+before transmission potentially reducing the overhead.  In general,
+larger peer tables results in lower overhead.  A configuration
+parameter controls the ratio of the size of the peer table storage to the
+primary table.  Note that the peer tables do not need to all be the same size and
+can shared a dynamic size to better optimize non-uniform data patterns.
+
+At end-of-stream, a worker flushes its caches and streams back its
+table (perhaps mixing in any spills) to its parent in response to the
+REST call that invoked that worker.  This process of
+writing results may block if the parent is not yet done.  If so, when the
+parent is done and ready, it will read the results from all of its children,
+and propagate the merged results in sorted order to its parent and so forth.
+
+The state required for reach worker to carry all this out is simply the list
+of endpoints that exist and any point in time along with the key space
+partitioning strategy.  Because of the properties of decomposable aggregations,
```

Comment:
```suggestion
partitioning strategy.  Because of the properties of composable aggregations,
```

Reply:
We can change the name but I got this term here: 
https://arxiv.org/pdf/1110.0725.pdf